### PR TITLE
Add flat object blog

### LINKED ISSUE #1525: [Blog] Flat Object in OpenSearch
**Describe the blog post your would like to write**
we are introducing  a new field type in OpenSearch: flat object. 

**What is the title of the blog post?**
Flat_Object field type in OpenSearch

**Who are the authors?**
Mingshi Liu

**What is the proposed posting date?**
6/13/23

### PR DESCRIPTION
**Pull request by kolchfa-aws**
### Description
Add flat object blog
 
### Issues Resolved
Fixes #1525 

### Check List
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the BSD-3-Clause License.

### CODE REVIEW

File: _posts/2023-06-09-flat-object.md [hunk 1]
```diff
@@ -0,0 +1,365 @@
...
+---
+
+OpenSearch 2.7 introduced the new `flat_object` field type. This field type is useful for objects with a large number of fields or when you are not familiar with the field names in your documents. The `flat_object` field type treats the entire JSON object as a string. Subfields within the JSON object are accessible using the flat object field name or the standard dot path notation, but they are not indexed for fast lookup. In this post, we explore how flat object simplifies mapping data structures and enhances the search experience in OpenSearch. 
+
+## Dynamic mapping
+
+In OpenSearch, a _mapping_ defines the structure of your data. It specifies field names, types, and indexing and analysis settings, ensuring that your data is organized and interpreted correctly. If you don't specify a custom mapping, OpenSearch infers the structure of your document automatically when you upload the document. This process is called _dynamic mapping_, where OpenSearch detects the document data structure and generates the corresponding mapping file.
```
Comment (mingshl) [inline _posts/2023-06-09-flat-object.md:20]: Shall we also link the mapping [doc](https://opensearch.org/docs/2.8/field-types/index/)?

**Review by mingshl — APPROVED** [review]
LGTM. Thanks @kolchfa-aws

**Comment by kolchfa-aws** [comment]
@vagimeli This blog is ready for editorial review. Thank you!

File: _posts/2023-06-09-flat-object.md [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
+        }
+      }
+    }
+  }
+}
+```
+
+### Searching with flat object
+
+When searching for a model parameter, you can the use `flat_object` field name, `Parameters`:
```
Comment (vagimeli) [inline _posts/2023-06-09-flat-object.md:289]:
```suggestion
When searching for a model parameter, you can the use the `flat_object` field name, `Parameters`: 
```

**Review by vagimeli — APPROVED** [review]
Editing complete. Minimal copyedits and comments.

File: _posts/2023-06-09-flat-object.md [hunk 3]
```diff
@@ -0,0 +1,365 @@
...
+layout: post
+title:  "Use flat object in OpenSearch"
+authors:
+- mingshl
+- kolchfa
+date: 2023-06-09
+categories:
+ - technical-post
+meta_keywords: 
+meta_description:
```
Comment (pajuric) [inline _posts/2023-06-09-flat-object.md:11]: Adding meta: 

**meta_keywords:** flat object field type, Mapping with flat object, OpenSearch mapping
**meta_description:** Explore how the flat object field type simplifies mapping data structures introduced in OpenSearch 2.7 and enhances the search experience.

File: _posts/2023-06-09-flat-object.md [hunk 4]
```diff
@@ -0,0 +1,365 @@
...
+categories:
+ - technical-post
+meta_keywords: 
+meta_description: 
+
+excerpt: OpenSearch 2.7 introduced the new flat object field type. This field type is useful for objects with a large number of fields or when you are not familiar with the field names in your documents. The flat object field type treats the entire JSON object as a string. Subfields within the JSON object are accessible using the flat object field name or the standard dot path notation, but they are not indexed for fast lookup. In this post, we explore how flat object simplifies mapping data structures and enhances the search experience in OpenSearch. 
+---
+
+OpenSearch 2.7 introduced the new `flat_object` field type. This field type is useful for objects with a large number of fields or when you are not familiar with the field names in your documents. The `flat_object` field type treats the entire JSON object as a string. Subfields within the JSON object are accessible using the `flat_object` field name or the standard dot path notation, but they are not indexed for fast lookup. In this post, we explore how flat object simplifies mapping data structures and enhances the search experience in OpenSearch.
```
Comment (pajuric) [inline _posts/2023-06-09-flat-object.md:16]: Consider linking to the OpenSearch 2.7 blog: https://opensearch.org/blog/get-started-opensearch-2-7-0/

File: _posts/2023-06-09-flat-object.md [hunk 5]
```diff
@@ -0,0 +1,365 @@
...
+---
+
+OpenSearch 2.7 introduced the new `flat_object` field type. This field type is useful for objects with a large number of fields or when you are not familiar with the field names in your documents. The `flat_object` field type treats the entire JSON object as a string. Subfields within the JSON object are accessible using the `flat_object` field name or the standard dot path notation, but they are not indexed for fast lookup. In this post, we explore how flat object simplifies mapping data structures and enhances the search experience in OpenSearch. 
+
+## Dynamic mapping
+
+In OpenSearch, a [_mapping_](https://opensearch.org/docs/latest/field-types/index/) defines the structure of your data. It specifies field names, types, and indexing and analysis settings, ensuring that your data is organized and interpreted correctly. If you don't specify a custom mapping, OpenSearch infers the structure of your document automatically when you upload the document. This process is called _dynamic mapping_, where OpenSearch detects the document data structure and generates the corresponding mapping file. 
+
+## When dynamic mapping falls flat
```
Comment (macohen) [inline _posts/2023-06-09-flat-object.md:22]: Hilarious! Love it.

File: _posts/2023-06-09-flat-object.md [hunk 6]
```diff
@@ -0,0 +1,365 @@
...
+
+When documents have complex data structures or deeply nested fields, relying on dynamic mapping can lead to the number of mapped fields in an index quickly growing to hundreds or even thousands. This "mapping explosion" negatively impacts the performance of your cluster. Additionally, searching through deeply nested indexes with lengthy dot paths can be inconvenient, especially if you are unfamiliar with the document structure. Flat object solves both of these problems.
```
Comment (macohen) [inline _posts/2023-06-09-flat-object.md:24]: can more detail be added about the negative impact of mapping explosions? What actually happens there? Are we able to reproduce a mapping explosion with this use case?

**Review by macohen — COMMENTED** [review]
helping the reader understand just how bad a mapping explosion can be would be a great addition to this blog post.

File: _posts/2023-06-09-flat-object.md (hunk 6 above)
Comment (mingshl) [reply _posts/2023-06-09-flat-object.md:24]: the example will be a mapping file that contains a lot of fields. But it maybe too long to add into the blog post. 

How about adding the error messages/symptoms  that you might get when you have mapping explosion issue? 

- OutOfMemoryError： If the mapping becomes too large to fit in memory, you may encounter an OutOfMemoryError, resulting in the unavailability of the cluster or nodes.
- MapperParsingException : If the number of unique field names grows too large, the cluster may throw exceptions such as MapperParsingException or IllegalArgumentException, indicating that the mapping update has failed.
- Performance degradation: As the mapping grows, the performance of indexing and search operations can degrade. Handling a large number of fields requires more resources and processing time, potentially leading to slower indexing and increased query latencies.
- Increased storage requirements: Each field in the mapping needs storage space. With a mapping explosion, the index's storage requirements can significantly increase, impacting disk space utilization and potentially leading to resource constraints.

**Comment by mingshl** [comment]
> @mingshl - If you can add the meta and update the date to 6/13/23, I can get this pushed live tomorrow. Please let me know when these items and all outstanding comments are addressed.
> 
> CC: @kolchfa-aws

Updated the date

File: _posts/2023-06-09-flat-object.md [hunk 7]
```diff
@@ -7,21 +7,30 @@ authors:
...
 
-When documents have complex data structures or deeply nested fields, relying on dynamic mapping can lead to the number of mapped fields in an index quickly growing to hundreds or even thousands. This "mapping explosion" negatively impacts the performance of your cluster. Additionally, searching through deeply nested indexes with lengthy dot paths can be inconvenient, especially if you are unfamiliar with the document structure. Flat object solves both of these problems.
+When documents have complex data structures or deeply nested fields, relying on dynamic mapping can lead to the number of mapped fields in an index quickly growing to hundreds or even thousands. This "mapping explosion" negatively impacts the performance of your cluster. 
+
+Symptoms of mapping explosion include:
+
+- `OutOfMemoryError`： If the mapping becomes too large to fit in memory, you may encounter an OutOfMemoryError, resulting in the unavailability of the cluster or nodes.
+- `MapperParsingException` : If the number of unique field names becomes too large, the cluster may throw exceptions such as `MapperParsingException` or `IllegalArgumentException`, indicating that the mapping update has failed.
```
Comment (vagimeli) [inline _posts/2023-06-09-flat-object.md:29]: Second bullet: Delete extra space between term and colon

**Review by vagimeli — APPROVED** [review]
LGTM

**Comment by kolchfa-aws** [comment]
@nateynateynate @pajuric This blog is ready to publish.

**Comment by pajuric** [comment]
@krisfreedain @nateynateynate - Blog is ready to push live today.

**Review by krisfreedain — APPROVED** [review]
LGTM!

**Comment by kolchfa-aws** [comment]
The meta is there @nateynateynate

**Comment by nateynateynate** [comment]
Yea that's my bad. Disregard.

**Pull request merged**